forui/lib/src/widgets/pagination/pagination_controller.dart

```diff
@@ -0,0 +1,147 @@
+import 'package:flutter/widgets.dart';
+
+import 'package:meta/meta.dart';
+
+import 'package:forui/forui.dart';
+
+/// A controller that controls which page is selected.
+class FPaginationController extends FChangeNotifier {
+  /// The number of sibling pages displayed beside the current page number. Defaults to 1.
+  ///
+  /// This value determines how many pages are shown on either side of the
+  /// currently selected page in the pagination.
+  ///
+  /// For example, if `siblings` is 2 and the current page is 5,
+  /// the displayed pages would be 3, 4, 5, 6, and 7.
+  ///
+  /// # Contract:
+  /// * Throws [AssertionError] if [siblings] < 0.
+  final int siblings;
+
+  /// Whether to show the first and last pages in the pagination. Defaults to `true`.
+  ///
+  /// If `true`, the pagination will always display the first and last page, regardless of the current page.
+  /// This can be useful for allowing users to quickly navigate to the beginning or end of the paginated content.
+  final bool showEdges;
+
+  /// A callback that is triggered whenever the current page changes.
+  ///
+  /// This callback is invoked whenever the `page` property of the `FPaginationController`
+  /// is updated, either programmatically or through user interaction.
+  ///
+  /// This can be useful for performing actions in response to page changes,
+  /// such as updating other parts of the UI or triggering analytics events.
+  final VoidCallback? onPageChanged;
+
+  final int _pages;
+  int _page;
+
+  /// Creates a [FPaginationController].
+  ///
+  /// # Contract:
+  /// * Throws [AssertionError] if 1 <= [initialPage] and [initialPage] <= length.
+  FPaginationController({
+    int pages = 1,
+    int initialPage = 0,
+    this.showEdges = true,
+    this.siblings = 1,
+    this.onPageChanged,
+  }) : assert(0 < pages, 'The total length of pages should be more than 0, but is $pages.'),
+       assert(0 <= siblings, 'The siblingLength should be non-negative, but is $siblings'),
+       assert(
+         0 <= initialPage && initialPage < pages,
+         'The initial page must be greater than or equal to 0 and less than or equal to length.',
+       ),
+       _page = initialPage,
+       _pages = pages;
+
+  /// Returns the total number of pages in the pagination.
+  int get pages => _pages;
+
+  /// Returns the current page index.
+  int get page => _page;
+
+  /// Moves to the previous page if the current page is greater than 1.
+  void previous() {
+    if (0 < _page) {
+      _page = _page - 1;
+      onPageChanged?.call();
+      notifyListeners();
+    }
+  }
+
+  /// Moves to the next page if the current page is less than the total number of pages.
+  void next() {
+    if (_page < pages - 1) {
+      _page = _page + 1;
+      onPageChanged?.call();
+      notifyListeners();
+    }
+  }
+
+  /// Sets the current page index.
+  set page(int index) {
+    if (0 <= index && index <= pages) {
```

Comment:
nitpick: I think it'll be much nicer if we test whether it is out of range. and then throw a StateError as a precondition.